# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [27]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [28]:
len(source_text)

9085267

In [29]:
source_text[:100]

'new jersey is sometimes quiet during autumn , and it is snowy in april .\nthe united states is usuall'

In [30]:
source_text[:100].split()

['new',
 'jersey',
 'is',
 'sometimes',
 'quiet',
 'during',
 'autumn',
 ',',
 'and',
 'it',
 'is',
 'snowy',
 'in',
 'april',
 '.',
 'the',
 'united',
 'states',
 'is',
 'usuall']

In [31]:
source_dict = {word: None for word in source_text.split()}

In [32]:
source_dict

{',': None,
 '.': None,
 '?': None,
 'a': None,
 'am': None,
 'and': None,
 'animal': None,
 'animals': None,
 'apple': None,
 'apple.': None,
 'apples': None,
 'apples.': None,
 'april': None,
 'are': None,
 "aren't": None,
 'august': None,
 'automobile': None,
 'autumn': None,
 'banana': None,
 'banana.': None,
 'bananas': None,
 'bananas.': None,
 'bear': None,
 'bears': None,
 'beautiful': None,
 'been': None,
 'between': None,
 'big': None,
 'bird': None,
 'birds': None,
 'black': None,
 'blue': None,
 'busy': None,
 'but': None,
 'california': None,
 'car': None,
 'cat': None,
 'cats': None,
 'chilly': None,
 'china': None,
 'chinese': None,
 'cold': None,
 'december': None,
 'did': None,
 "didn't": None,
 'difficult': None,
 'dislike': None,
 'disliked': None,
 'dislikes': None,
 'do': None,
 'does': None,
 'dog': None,
 'dogs': None,
 'drives': None,
 'driving': None,
 'drove': None,
 'dry': None,
 'during': None,
 'easy': None,
 'eiffel': None,
 'elephant': None,
 'elephants':

> **Note:** `{word: None for word in some_text.split()}` returns a dictionary with one entry for each unique token. Cute!

In [33]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [34]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # Implement Function
    # 1. split into sentences 2. transform to int 3. each sentence a list 4. EOS on target sentences
    source_text_sentences = [sentence for sentence in source_text.split('\n')]
    source_id_text = [[source_vocab_to_int.get(word, source_vocab_to_int['<UNK>']) for word in s.split()] for s in source_text_sentences]
    target_text_sentences = [sentence for sentence in target_text.split('\n')]
    target_id_text = [[target_vocab_to_int.get(word, target_vocab_to_int['<UNK>']) for word in s.split()]+[target_vocab_to_int['<EOS>']] for s in target_text_sentences]
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [35]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [38]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # Implement Function
    input_data =             tf.placeholder(tf.int32, [None, None], name='input')
    targets =                tf.placeholder(tf.int32, [None, None], name='targets')
    lr =                     tf.placeholder(tf.float32, name='learning_rate')
    keep_prob =              tf.placeholder(tf.float32, name='keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None,], name='target_sequence_length')
    max_target_len =         tf.reduce_max(target_sequence_length, name='max_target_len')
    source_sequence_length = tf.placeholder(tf.int32, [None,], name='source_sequence_length')
    
    return input_data, targets, lr, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "/usr/local/lib/python3.5/runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "/usr/local/lib/python3.5/runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "/usr/local/lib/python3.5/site-packages/ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "/usr/local/lib/python3.5/site-packages/traitlets/config/application.py", line 658, in launch_instance\n    app.start()', 'File "/usr/local/lib/python3.5/site-packages/ipykernel/kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "/usr/local/lib/python3.5/site-packages/zmq/eventloop/ioloop.py", line 177, in start\n    super(ZMQIOLoop, self).start()', 'F

Tests Passed


> **Note:** Found [this discussion](https://github.com/tensorflow/tensorflow/issues/11315) confirming that this is a _feature_ of TF 1.2.x and suggesting a best practice in this case. It doesn't seem to be a problem with environment corruption as suggested in the [Udacity forum](https://discussions.udacity.com/t/model-inputs-throws-tensorflow-errors/286123/9). The error is clearly thrown by the unit test.

### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [39]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # Implement Function
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)

    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [40]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # Implement Function
    # Encoder embedding
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    # RNN cell
    def build_cell(rnn_size):
        lstm = tf.contrib.rnn.LSTMCell(rnn_size, initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
        return drop
    
    rnn_cell = tf.contrib.rnn.MultiRNNCell([build_cell(rnn_size) for _ in range(num_layers)])
    
    rnn_output, rnn_state = tf.nn.dynamic_rnn(rnn_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return rnn_output, rnn_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [41]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # Implement Function
    # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input,
                                                        sequence_length=target_sequence_length,
                                                        time_major=False)
        
        
    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                       training_helper,
                                                       encoder_state,
                                                       output_layer) 
        
    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder,
                                                                impute_finished=True,
                                                                maximum_iterations=max_summary_length)[0]

    
    
    return training_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


> **Note:** The `keep_prob` parameter, which this [Udacity forum answer](https://discussions.udacity.com/t/decoding-layer-train-keep-prob/242022) suggests can be applied to the outputs of [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder), seems to be a leftover from a previous iteration. The said function is part of the TF 1.0 API and is deprecated in later versions. In the current implementation method using `TrainingHelper`, `BasicDecoder`, and `dynamic_decode`, there doesn't seem to be an obvious place to apply dropout.

### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [42]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # Implement Function
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Helper for the inference process.
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings,
                                                                start_tokens,
                                                                end_of_sequence_id)

    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell,
                                                        inference_helper,
                                                        encoder_state,
                                                        output_layer)
        
    # Perform dynamic decoding using the decoder
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder,
                                                                 impute_finished=True,
                                                                 maximum_iterations=max_target_sequence_length)[0]

    
    return inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


> **Note:** There doesn't seem to be any use for the `vocab_size` parameter in any of the functions used.

### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [43]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Implement Function
    # 1. Decoder Embedding
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        dec_cell = tf.contrib.rnn.DropoutWrapper(dec_cell, output_keep_prob=keep_prob)
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    # 4. Training decoder
    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(encoder_state, 
                                                       dec_cell, 
                                                       dec_embed_input, 
                                                       target_sequence_length, 
                                                       max_target_sequence_length, 
                                                       output_layer, 
                                                       keep_prob)

    # 5. Inference decoder
    with tf.variable_scope("decode", reuse=True):
        inference_decoder_output = decoding_layer_infer(encoder_state, 
                                                       dec_cell, 
                                                       dec_embeddings, 
                                                       target_vocab_to_int['<GO>'], 
                                                       target_vocab_to_int['<EOS>'], 
                                                       max_target_sequence_length, 
                                                       target_vocab_size, 
                                                       output_layer, 
                                                       batch_size, 
                                                       keep_prob)
        
        
    
    return training_decoder_output, inference_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [44]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Implement Function
    # 1. Pass the input data through the encoder. We'll ignore the encoder output, but use the state
    _, enc_state = encoding_layer(input_data, 
                                  rnn_size, 
                                  num_layers, 
                                  keep_prob,  
                                  source_sequence_length, 
                                  source_vocab_size, 
                                  enc_embedding_size)    
    
    # 2. Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # 3. Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, 
                                                                      enc_state, 
                                                                      target_sequence_length, 
                                                                      max_target_sentence_length, 
                                                                      rnn_size, 
                                                                      num_layers, 
                                                                      target_vocab_to_int, 
                                                                      target_vocab_size, 
                                                                      batch_size, 
                                                                      keep_prob, 
                                                                      dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [4]:
# Number of Epochs
epochs = 25
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 100
decoding_embedding_size = 100
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.75
display_step = batch_size // 4

### Build the Graph
Build the graph using the neural network you implemented.

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# save_path = 'checkpoints/dev'
save_path = '/output/checkpoints/dev' # For running on floydhub
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   32/1077 - Train Accuracy: 0.4461, Validation Accuracy: 0.4588, Loss: 2.5979
Epoch   0 Batch   64/1077 - Train Accuracy: 0.4262, Validation Accuracy: 0.4961, Loss: 2.2756
Epoch   0 Batch   96/1077 - Train Accuracy: 0.4352, Validation Accuracy: 0.4865, Loss: 1.8501
Epoch   0 Batch  128/1077 - Train Accuracy: 0.4427, Validation Accuracy: 0.4599, Loss: 1.4981
Epoch   0 Batch  160/1077 - Train Accuracy: 0.4707, Validation Accuracy: 0.5028, Loss: 1.3920
Epoch   0 Batch  192/1077 - Train Accuracy: 0.4695, Validation Accuracy: 0.4890, Loss: 1.2841
Epoch   0 Batch  224/1077 - Train Accuracy: 0.4934, Validation Accuracy: 0.5511, Loss: 1.1576
Epoch   0 Batch  256/1077 - Train Accuracy: 0.5074, Validation Accuracy: 0.5643, Loss: 1.0797
Epoch   0 Batch  288/1077 - Train Accuracy: 0.5117, Validation Accuracy: 0.5831, Loss: 0.9767
Epoch   0 Batch  320/1077 - Train Accuracy: 0.5969, Validation Accuracy: 0.5955, Loss: 0.8873
Epoch   0 Batch  352/1077 - Train Accuracy: 0.5488, Validati

Epoch   2 Batch  736/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9325, Loss: 0.0355
Epoch   2 Batch  768/1077 - Train Accuracy: 0.9379, Validation Accuracy: 0.9432, Loss: 0.0465
Epoch   2 Batch  800/1077 - Train Accuracy: 0.9590, Validation Accuracy: 0.9595, Loss: 0.0362
Epoch   2 Batch  832/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9304, Loss: 0.0340
Epoch   2 Batch  864/1077 - Train Accuracy: 0.9371, Validation Accuracy: 0.9403, Loss: 0.0405
Epoch   2 Batch  896/1077 - Train Accuracy: 0.9622, Validation Accuracy: 0.9496, Loss: 0.0425
Epoch   2 Batch  928/1077 - Train Accuracy: 0.9641, Validation Accuracy: 0.9503, Loss: 0.0366
Epoch   2 Batch  960/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9435, Loss: 0.0409
Epoch   2 Batch  992/1077 - Train Accuracy: 0.9586, Validation Accuracy: 0.9549, Loss: 0.0448
Epoch   2 Batch 1024/1077 - Train Accuracy: 0.9309, Validation Accuracy: 0.9485, Loss: 0.0487
Epoch   2 Batch 1056/1077 - Train Accuracy: 0.9652, Validati

Epoch   5 Batch  384/1077 - Train Accuracy: 0.9910, Validation Accuracy: 0.9709, Loss: 0.0197
Epoch   5 Batch  416/1077 - Train Accuracy: 0.9762, Validation Accuracy: 0.9588, Loss: 0.0241
Epoch   5 Batch  448/1077 - Train Accuracy: 0.9730, Validation Accuracy: 0.9624, Loss: 0.0229
Epoch   5 Batch  480/1077 - Train Accuracy: 0.9885, Validation Accuracy: 0.9560, Loss: 0.0146
Epoch   5 Batch  512/1077 - Train Accuracy: 0.9863, Validation Accuracy: 0.9759, Loss: 0.0168
Epoch   5 Batch  544/1077 - Train Accuracy: 0.9727, Validation Accuracy: 0.9684, Loss: 0.0132
Epoch   5 Batch  576/1077 - Train Accuracy: 0.9799, Validation Accuracy: 0.9759, Loss: 0.0191
Epoch   5 Batch  608/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9947, Loss: 0.0220
Epoch   5 Batch  640/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9755, Loss: 0.0193
Epoch   5 Batch  672/1077 - Train Accuracy: 0.9829, Validation Accuracy: 0.9606, Loss: 0.0149
Epoch   5 Batch  704/1077 - Train Accuracy: 0.9598, Validati

Epoch   8 Batch   32/1077 - Train Accuracy: 0.9874, Validation Accuracy: 0.9805, Loss: 0.0091
Epoch   8 Batch   64/1077 - Train Accuracy: 0.9930, Validation Accuracy: 0.9705, Loss: 0.0085
Epoch   8 Batch   96/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9783, Loss: 0.0179
Epoch   8 Batch  128/1077 - Train Accuracy: 0.9896, Validation Accuracy: 0.9826, Loss: 0.0094
Epoch   8 Batch  160/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9822, Loss: 0.0129
Epoch   8 Batch  192/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9691, Loss: 0.0135
Epoch   8 Batch  224/1077 - Train Accuracy: 0.9805, Validation Accuracy: 0.9794, Loss: 0.0151
Epoch   8 Batch  256/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9822, Loss: 0.0199
Epoch   8 Batch  288/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9833, Loss: 0.0192
Epoch   8 Batch  320/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9773, Loss: 0.0163
Epoch   8 Batch  352/1077 - Train Accuracy: 0.9898, Validati

Epoch  10 Batch  736/1077 - Train Accuracy: 0.9922, Validation Accuracy: 0.9869, Loss: 0.0106
Epoch  10 Batch  768/1077 - Train Accuracy: 0.9926, Validation Accuracy: 0.9862, Loss: 0.0079
Epoch  10 Batch  800/1077 - Train Accuracy: 0.9941, Validation Accuracy: 0.9922, Loss: 0.0090
Epoch  10 Batch  832/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9876, Loss: 0.0097
Epoch  10 Batch  864/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9879, Loss: 0.0109
Epoch  10 Batch  896/1077 - Train Accuracy: 0.9959, Validation Accuracy: 0.9897, Loss: 0.0097
Epoch  10 Batch  928/1077 - Train Accuracy: 0.9926, Validation Accuracy: 0.9805, Loss: 0.0111
Epoch  10 Batch  960/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9815, Loss: 0.0117
Epoch  10 Batch  992/1077 - Train Accuracy: 0.9910, Validation Accuracy: 0.9929, Loss: 0.0122
Epoch  10 Batch 1024/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9854, Loss: 0.0129
Epoch  10 Batch 1056/1077 - Train Accuracy: 0.9910, Validati

Epoch  13 Batch  416/1077 - Train Accuracy: 0.9879, Validation Accuracy: 0.9790, Loss: 0.0085
Epoch  13 Batch  448/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9790, Loss: 0.0096
Epoch  13 Batch  480/1077 - Train Accuracy: 0.9975, Validation Accuracy: 0.9790, Loss: 0.0086
Epoch  13 Batch  512/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9787, Loss: 0.0057
Epoch  13 Batch  544/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9783, Loss: 0.0054
Epoch  13 Batch  576/1077 - Train Accuracy: 0.9877, Validation Accuracy: 0.9854, Loss: 0.0061
Epoch  13 Batch  608/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9794, Loss: 0.0073
Epoch  13 Batch  640/1077 - Train Accuracy: 0.9996, Validation Accuracy: 0.9790, Loss: 0.0055
Epoch  13 Batch  672/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9801, Loss: 0.0078
Epoch  13 Batch  704/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9805, Loss: 0.0118
Epoch  13 Batch  736/1077 - Train Accuracy: 0.9819, Validati

Epoch  16 Batch   64/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9883, Loss: 0.0074
Epoch  16 Batch   96/1077 - Train Accuracy: 0.9930, Validation Accuracy: 0.9819, Loss: 0.0086
Epoch  16 Batch  128/1077 - Train Accuracy: 0.9978, Validation Accuracy: 0.9883, Loss: 0.0044
Epoch  16 Batch  160/1077 - Train Accuracy: 0.9898, Validation Accuracy: 0.9883, Loss: 0.0053
Epoch  16 Batch  192/1077 - Train Accuracy: 0.9953, Validation Accuracy: 0.9879, Loss: 0.0075
Epoch  16 Batch  224/1077 - Train Accuracy: 0.9934, Validation Accuracy: 0.9830, Loss: 0.0074
Epoch  16 Batch  256/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9858, Loss: 0.0111
Epoch  16 Batch  288/1077 - Train Accuracy: 0.9891, Validation Accuracy: 0.9833, Loss: 0.0115
Epoch  16 Batch  320/1077 - Train Accuracy: 0.9977, Validation Accuracy: 0.9822, Loss: 0.0033
Epoch  16 Batch  352/1077 - Train Accuracy: 0.9934, Validation Accuracy: 0.9833, Loss: 0.0036
Epoch  16 Batch  384/1077 - Train Accuracy: 0.9992, Validati

Epoch  18 Batch  768/1077 - Train Accuracy: 0.9945, Validation Accuracy: 0.9748, Loss: 0.0035
Epoch  18 Batch  800/1077 - Train Accuracy: 0.9945, Validation Accuracy: 0.9737, Loss: 0.0026
Epoch  18 Batch  832/1077 - Train Accuracy: 0.9898, Validation Accuracy: 0.9641, Loss: 0.0044
Epoch  18 Batch  864/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9854, Loss: 0.0057
Epoch  18 Batch  896/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9723, Loss: 0.0059
Epoch  18 Batch  928/1077 - Train Accuracy: 0.9992, Validation Accuracy: 0.9751, Loss: 0.0037
Epoch  18 Batch  960/1077 - Train Accuracy: 0.9948, Validation Accuracy: 0.9748, Loss: 0.0041
Epoch  18 Batch  992/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9790, Loss: 0.0048
Epoch  18 Batch 1024/1077 - Train Accuracy: 0.9930, Validation Accuracy: 0.9638, Loss: 0.0061
Epoch  18 Batch 1056/1077 - Train Accuracy: 0.9891, Validation Accuracy: 0.9716, Loss: 0.0050
Epoch  19 Batch   32/1077 - Train Accuracy: 0.9862, Validati

Epoch  21 Batch  416/1077 - Train Accuracy: 0.9906, Validation Accuracy: 0.9826, Loss: 0.0063
Epoch  21 Batch  448/1077 - Train Accuracy: 0.9980, Validation Accuracy: 0.9801, Loss: 0.0043
Epoch  21 Batch  480/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9751, Loss: 0.0083
Epoch  21 Batch  512/1077 - Train Accuracy: 0.9895, Validation Accuracy: 0.9755, Loss: 0.0040
Epoch  21 Batch  544/1077 - Train Accuracy: 0.9996, Validation Accuracy: 0.9755, Loss: 0.0063
Epoch  21 Batch  576/1077 - Train Accuracy: 0.9934, Validation Accuracy: 0.9734, Loss: 0.0044
Epoch  21 Batch  608/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9783, Loss: 0.0067
Epoch  21 Batch  640/1077 - Train Accuracy: 0.9948, Validation Accuracy: 0.9805, Loss: 0.0046
Epoch  21 Batch  672/1077 - Train Accuracy: 0.9996, Validation Accuracy: 0.9869, Loss: 0.0038
Epoch  21 Batch  704/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9780, Loss: 0.0078
Epoch  21 Batch  736/1077 - Train Accuracy: 0.9996, Validati

Epoch  24 Batch   64/1077 - Train Accuracy: 0.9988, Validation Accuracy: 0.9783, Loss: 0.0047
Epoch  24 Batch   96/1077 - Train Accuracy: 1.0000, Validation Accuracy: 0.9833, Loss: 0.0029
Epoch  24 Batch  128/1077 - Train Accuracy: 0.9978, Validation Accuracy: 0.9904, Loss: 0.0019
Epoch  24 Batch  160/1077 - Train Accuracy: 0.9949, Validation Accuracy: 0.9879, Loss: 0.0020
Epoch  24 Batch  192/1077 - Train Accuracy: 0.9910, Validation Accuracy: 0.9858, Loss: 0.0076
Epoch  24 Batch  224/1077 - Train Accuracy: 0.9996, Validation Accuracy: 0.9925, Loss: 0.0039
Epoch  24 Batch  256/1077 - Train Accuracy: 0.9973, Validation Accuracy: 0.9876, Loss: 0.0061
Epoch  24 Batch  288/1077 - Train Accuracy: 0.9992, Validation Accuracy: 0.9876, Loss: 0.0091
Epoch  24 Batch  320/1077 - Train Accuracy: 0.9910, Validation Accuracy: 0.9755, Loss: 0.0062
Epoch  24 Batch  352/1077 - Train Accuracy: 0.9945, Validation Accuracy: 0.9751, Loss: 0.0039
Epoch  24 Batch  384/1077 - Train Accuracy: 1.0000, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [2]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # Implement Function
    
    return [vocab_to_int.get(word, vocab_to_int['<UNK>']) for word in sentence.lower().split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [5]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from /output/checkpoints/dev
Input
  Word Ids:      [138, 219, 200, 111, 46, 59, 218]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [282, 230, 260, 113, 4, 14, 96, 61, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


> **Brief exploration:** It would be interesting to come up with a few sentences based on the 227 English words and see how well the model generalizes.

In [6]:
len(source_vocab_to_int)

231

In [11]:
print(sorted([word for word, _ in source_vocab_to_int.items()]))

[',', '.', '<EOS>', '<GO>', '<PAD>', '<UNK>', '?', 'a', 'am', 'and', 'animal', 'animals', 'apple', 'apple.', 'apples', 'apples.', 'april', 'are', "aren't", 'august', 'automobile', 'autumn', 'banana', 'banana.', 'bananas', 'bananas.', 'bear', 'bears', 'beautiful', 'been', 'between', 'big', 'bird', 'birds', 'black', 'blue', 'busy', 'but', 'california', 'car', 'cat', 'cats', 'chilly', 'china', 'chinese', 'cold', 'december', 'did', "didn't", 'difficult', 'dislike', 'disliked', 'dislikes', 'do', 'does', 'dog', 'dogs', 'drives', 'driving', 'drove', 'dry', 'during', 'easy', 'eiffel', 'elephant', 'elephants', 'english', 'fall', 'favorite', 'favorite.', 'feared', 'february', 'field', 'football', 'france', 'freezing', 'french', 'fruit', 'fruit.', 'fun', 'go', 'going', 'grape', 'grape.', 'grapefruit', 'grapefruit.', 'grapes', 'grapes.', 'green', 'grocery', 'has', 'have', 'he', 'her', 'his', 'horse', 'horses', 'hot', 'how', 'i', 'in', 'india', 'is', "isn't", 'it', "it's", 'january', 'jersey', 'jul

In [13]:
sentences = [
    'monkeys dislike snakes .',
    'my favorite fruit isn\'t the mango .',
    'how wet is it in paris in november ?',
    'beautiful animals are driving a automobile .',
    'a warm winter is nice .',
    'do you think california has a dry winter ?',
    'the eiffel tower was rusty last july .',
    'it is difficult to translate from chineses to spanish .',
    'elephants , mice , and horses loved football .',
    'i usually like a big yellow grapefruit when i am hot .'
            ]

In [14]:
loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    for translate_sentence in sentences:
        translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)
        
        input_data = loaded_graph.get_tensor_by_name('input:0')
        logits = loaded_graph.get_tensor_by_name('predictions:0')
        target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
        source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
        keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

        translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                             target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                             source_sequence_length: [len(translate_sentence)]*batch_size,
                                             keep_prob: 1.0})[0]

        print('Input')
        print('  Word Ids:      {}'.format([i for i in translate_sentence]))
        print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

        print('\nPrediction')
        print('  Word Ids:      {}'.format([i for i in translate_logits]))
        print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from /output/checkpoints/dev
Input
  Word Ids:      [112, 143, 131, 218]
  English Words: ['monkeys', 'dislike', 'snakes', '.']

Prediction
  Word Ids:      [16, 94, 91, 259, 120, 350, 61, 1]
  French Words: singes est verts et les raisins . <EOS>
Input
  Word Ids:      [183, 8, 126, 166, 179, 204, 218]
  English Words: ['my', 'favorite', 'fruit', "isn't", 'the', 'mango', '.']

Prediction
  Word Ids:      [65, 197, 277, 94, 150, 219, 61, 1]
  French Words: mon fruit préféré est la mangue . <EOS>
Input
  Word Ids:      [196, 206, 164, 124, 71, 195, 71, 35, 72]
  English Words: ['how', 'wet', 'is', 'it', 'in', 'paris', 'in', 'november', '?']

Prediction
  Word Ids:      [77, 94, 207, 99, 173, 45, 28, 337, 282, 94, 336, 173, 242, 61, 1]
  French Words: new est le gel en novembre , mais il est pluvieux en été . <EOS>
Input
  Word Ids:      [62, 109, 146, 153, 200, 190, 218]
  English Words: ['beautiful', 'animals', 'are', 'driving', 'a', 'automobile', '

> **Final comments:** The almost infinite compositionality of language easily defeats such a toy model. It obviously doen't generalize well on arbitrary sentences using the words in its vocabulary.

## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.